In [1]:
import pandas as pd
from ast import literal_eval


In [75]:
pd.set_option('display.max_rows',1000)

In [6]:
df = pd.read_csv("/Users/ishitagopal/Downloads/the-movies-dataset/credits.csv")

In [7]:
df.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

In [8]:
df.shape

(45476, 3)

In [9]:
# example -- exploring one is entry in the cast column
# 'name' contains the actors real name
literal_eval(df['cast'][1])

[{'cast_id': 1,
  'character': 'Alan Parrish',
  'credit_id': '52fe44bfc3a36847f80a7c73',
  'gender': 2,
  'id': 2157,
  'name': 'Robin Williams',
  'order': 0,
  'profile_path': '/sojtJyIV3lkUeThD7A2oHNm8183.jpg'},
 {'cast_id': 8,
  'character': 'Samuel Alan Parrish / Van Pelt',
  'credit_id': '52fe44bfc3a36847f80a7c99',
  'gender': 2,
  'id': 8537,
  'name': 'Jonathan Hyde',
  'order': 1,
  'profile_path': '/7il5D76vx6QVRVlpVvBPEC40MBi.jpg'},
 {'cast_id': 2,
  'character': 'Judy Sheperd',
  'credit_id': '52fe44bfc3a36847f80a7c77',
  'gender': 1,
  'id': 205,
  'name': 'Kirsten Dunst',
  'order': 2,
  'profile_path': '/wBXvh6PJd0IUVNpvatPC1kzuHtm.jpg'},
 {'cast_id': 24,
  'character': 'Peter Shepherd',
  'credit_id': '52fe44c0c3a36847f80a7ce7',
  'gender': 0,
  'id': 145151,
  'name': 'Bradley Pierce',
  'order': 3,
  'profile_path': '/j6iW0vVA23GQniAPSYI6mi4hiEW.jpg'},
 {'cast_id': 10,
  'character': 'Sarah Whittle',
  'credit_id': '52fe44bfc3a36847f80a7c9d',
  'gender': 1,
  'id': 5

In [10]:
# Extracting the names of the first 5 actors in each movie
# Create a list of lists
actors = []

for index, row in df.iterrows():
    five = []
    for item in literal_eval(df['cast'][index])[0:5]:
        five.append(item["name"])
    actors.append(five)

In [11]:
# Example output for the first three 5 rows in df
actors[0:5]

[['Tom Hanks', 'Tim Allen', 'Don Rickles', 'Jim Varney', 'Wallace Shawn'],
 ['Robin Williams',
  'Jonathan Hyde',
  'Kirsten Dunst',
  'Bradley Pierce',
  'Bonnie Hunt'],
 ['Walter Matthau',
  'Jack Lemmon',
  'Ann-Margret',
  'Sophia Loren',
  'Daryl Hannah'],
 ['Whitney Houston',
  'Angela Bassett',
  'Loretta Devine',
  'Lela Rochon',
  'Gregory Hines'],
 ['Steve Martin',
  'Diane Keaton',
  'Martin Short',
  'Kimberly Williams-Paisley',
  'George Newbern']]

In [13]:
# check if legths are equal
len(actors)==len(df)

# add the above list as a column to df -- 'first5actors'
df['first5actors'] = actors

In [14]:
df.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602   
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357   
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862   

                                        first5actors  
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst,...  
2  [Walter Matthau, Jack Lemmon, Ann-Margret, Sop...  
3  [Whitney Houston, Angela Bassett, Loretta Devi...  
4  [Steve Martin, Diane Keaton, Martin Short, Kim...

In [15]:
# Attach the movie's year of release 
df2 = pd.read_csv("/Users/ishitagopal/Downloads/the-movies-dataset/movies_metadata.csv")

In [16]:
df2.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [17]:
# data types for id column were different -- wasnt merging -- converted both to str because int was giving errors
df['id'] = df['id'].astype('str')
df2['id'] = df2['id'].astype('str')

In [18]:
# Merging df and df2 on the column id 
# Restricting columns

combined = df[['id','first5actors']].merge(df2[['id','imdb_id','release_date']],on='id')


In [19]:
combined.shape


(45538, 4)

In [20]:
combined.head()

id                                       first5actors    imdb_id  \
0    862  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  tt0114709   
1   8844  [Robin Williams, Jonathan Hyde, Kirsten Dunst,...  tt0113497   
2  15602  [Walter Matthau, Jack Lemmon, Ann-Margret, Sop...  tt0113228   
3  31357  [Whitney Houston, Angela Bassett, Loretta Devi...  tt0114885   
4  11862  [Steve Martin, Diane Keaton, Martin Short, Kim...  tt0113041   

  release_date  
0   1995-10-30  
1   1995-12-15  
2   1995-12-22  
3   1995-12-22  
4   1995-02-10

In [21]:
# read in imdb dataset that has the birth years for actors

tsv_file="/Users/ishitagopal/Downloads/name.basics.tsv"
imdb_csv=pd.read_table(tsv_file,sep='\t')



In [22]:
imdb_csv.head()

nconst      primaryName birthYear deathYear  \
0  nm0000001     Fred Astaire      1899      1987   
1  nm0000002    Lauren Bacall      1924      2014   
2  nm0000003  Brigitte Bardot      1934        \N   
3  nm0000004     John Belushi      1949      1982   
4  nm0000005   Ingmar Bergman      1918      2007   

                primaryProfession                           knownForTitles  
0  soundtrack,actor,miscellaneous  tt0053137,tt0043044,tt0050419,tt0072308  
1              actress,soundtrack  tt0037382,tt0071877,tt0038355,tt0117057  
2     actress,soundtrack,producer  tt0057345,tt0054452,tt0059956,tt0049189  
3         actor,writer,soundtrack  tt0072562,tt0077975,tt0078723,tt0080455  
4           writer,director,actor  tt0083922,tt0050986,tt0069467,tt0050976

In [23]:
# Create two lists -- name of actor -- & -- corresponding birth yr
actor_names = imdb_csv['primaryName'].tolist()
actor_birth_yrs = imdb_csv['birthYear'].tolist()

In [24]:
# example -- way to match name with birthyr
name = "Tom Hanks"
if name in actor_names:
    print(actor_names.index(name))
    print()

157



In [66]:
# Applying the above logic to the entire dataset 
# This takes a while
combined_birthyr5 = []

for index, row in combined.iterrows():
    birthyr5 = []
    for name in row['first5actors']:
        if name in actor_names:
            birthyr5.append(actor_birth_yrs[actor_names.index(name)])
        else:
            birthyr5.append('no record')
    combined_birthyr5.append(birthyr5)
        
    

In [69]:
len(combined_birthyr5)==len(combined)

True

In [70]:
combined["birthyr5"]=combined_birthyr5

In [71]:
combined.head()

id                                       first5actors    imdb_id  \
0    862  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  tt0114709   
1   8844  [Robin Williams, Jonathan Hyde, Kirsten Dunst,...  tt0113497   
2  15602  [Walter Matthau, Jack Lemmon, Ann-Margret, Sop...  tt0113228   
3  31357  [Whitney Houston, Angela Bassett, Loretta Devi...  tt0114885   
4  11862  [Steve Martin, Diane Keaton, Martin Short, Kim...  tt0113041   

  release_date                        birthyr5  
0   1995-10-30  [1956, 1953, 1926, 1949, 1943]  
1   1995-12-15  [1951, 1948, 1982, 1982, 1961]  
2   1995-12-22  [1920, 1925, 1941, 1934, 1960]  
3   1995-12-22  [1963, 1958, 1949, 1964, 1946]  
4   1995-02-10  [1945, 1946, 1950, 1971, 1964]

In [72]:
# Write resulting dataframe to csv file 
combined.to_csv(r"/Users/ishitagopal/Downloads/combined.csv")

In [78]:
# Checking if everything was saved correctly

load = pd.read_csv("/Users/ishitagopal/Downloads/combined.csv")

In [80]:
load['birthyr5']

0            ['1956', '1953', '1926', '1949', '1943']
1            ['1951', '1948', '1982', '1982', '1961']
2            ['1920', '1925', '1941', '1934', '1960']
3            ['1963', '1958', '1949', '1964', '1946']
4            ['1945', '1946', '1950', '1971', '1964']
                             ...                     
45533                          ['1972', '\\N', '\\N']
45534         ['1973', '\\N', '1986', '1961', '1963']
45535        ['1969', '1962', '1973', '1953', '1963']
45536    ['no record', '1886', '\\N', '1888', '1918']
45537                                              []
Name: birthyr5, Length: 45538, dtype: object

In [86]:
load.shape

(45538, 6)

### Notes on values in the birthyr5 column
1) '\\N' -- this is how the year is saved in the imdb dataset 
2) 'no record' -- this indicates that the actors identified in df (credits.csv) were not found in the imdb dataset`
3) where column contains less than 5 values -- it means the actors identified in df (credits.csv) were less than 5
   example: id 45533 has only 3 actors